### ISSUE TITLE
Heka with the AMQP input plugin shuts down when the connection to the AMQP server drops

### ISSUE THREAD

**Issue by simonpasquier** [mozilla-services/heka#1757]
We've got Heka configured to read data from AMQP (RabbitMQ). With version 0.9, Heka was able to cope when the connection to the server was dropped but it looks like something is broken with 0.10.

Here is what happens when the AMQP server becomes unavailable:

```
2015/10/08 11:15:03 Pre-loading: [openstack_info_amqp]
2015/10/08 11:15:03 Pre-loading: [notification_decoder]
2015/10/08 11:15:03 Pre-loading: [HekaFramingSplitter]
2015/10/08 11:15:03 Loading: [HekaFramingSplitter]
2015/10/08 11:15:03 Pre-loading: [NullSplitter]
2015/10/08 11:15:03 Loading: [NullSplitter]
2015/10/08 11:15:03 Pre-loading: [ProtobufDecoder]
2015/10/08 11:15:03 Loading: [ProtobufDecoder]
2015/10/08 11:15:03 Pre-loading: [ProtobufEncoder]
2015/10/08 11:15:03 Loading: [ProtobufEncoder]
2015/10/08 11:15:03 Pre-loading: [TokenSplitter]
2015/10/08 11:15:03 Loading: [TokenSplitter]
2015/10/08 11:15:03 Loading: [notification_decoder]
2015/10/08 11:15:03 Loading: [openstack_info_amqp]
2015/10/08 11:15:03 Starting hekad...
2015/10/08 11:15:03 MessageRouter started.
2015/10/08 11:15:03 Input started: openstack_info_amqp
2015/10/08 11:21:06 Shutdown initiated.
2015/10/08 11:21:06 Stop message sent to input 'openstack_info_amqp'
2015/10/08 11:21:06 Waiting for decoders shutdown
2015/10/08 11:21:06 Stop message sent to decoder 'openstack_info_amqp-notification_decoder'
2015/10/08 11:21:06 Decoder 'openstack_info_amqp-notification_decoder': stopped
2015/10/08 11:21:06 Decoders shutdown complete
2015/10/08 11:21:06 Shutdown complete.
```

As @zstyblik noted in #1756, I suspect that this is more related to the AMQP library because the amqp_input.go code hasn't changed a lot between 0.9 and 0.10.

**Comment by simonpasquier** [comment]
IIUC the Run() method should return something other than nil [1] to trigger a restart of the plugin. Otherwise the pipeline runner triggers a shutdown of Heka.

[1] https://github.com/mozilla-services/heka/blob/1af899d0b975893bbf19d3bcb8c2e6653a7c9252/plugins/amqp/amqp_input.go#L206

**Comment by rafrombrc** [comment]
This should be resolved w/ the merging of #1764 and #1794.
